system/updated/casync/tar.py

```diff
@@ -0,0 +1,40 @@
+import pathlib
+import tarfile
+from typing import IO, Callable
+
+
+def include_default(_) -> bool:
+  return True
+
+
+def create_tar_archive(fh: IO[bytes], directory: pathlib.Path, include: Callable[[pathlib.Path], bool] = include_default):
+  """Creates a tar archive of a directory"""
+
+  tar = tarfile.open(fileobj=fh, mode='w')
+  for file in sorted(directory.rglob("*"), key=lambda f: f.stat().st_size if f.is_file() else 0, reverse=True):
+    if not include(file):
+      continue
+    relative_path = str(file.relative_to(directory))
+    if file.is_symlink():
+      info = tarfile.TarInfo(relative_path)
+      info.type = tarfile.SYMTYPE
+      info.linkpath = str(file.readlink())
+      tar.addfile(info)
+
+    elif file.is_file():
+      info = tarfile.TarInfo(relative_path)
+      info.size = file.stat().st_size
+      info.type = tarfile.REGTYPE
+      with file.open('rb') as f:
+        tar.addfile(info, f)
+
+  tar.close()
+  fh.seek(0)
```

Comment:
what's this for? seems odd to pass in a file handle anyway

Reply:
thought we might want to do this in memory eventually but I don't think that will be the case.